resolwe/flow/filters.py

```diff
@@ -337,6 +339,63 @@ class EntityFilter(BaseCollectionFilter):
     """Filter the Entity endpoint."""
 
     relation_id = filters.NumberFilter(field_name="relation__id")
+    annotations = filters.CharFilter(method="filter_annotations")
+
+    def filter_annotations(self, queryset, name, value):
+        """Filter entities by annotations.
+
+        The value must be in the following format: field_id__lookup_type:value
+        """
+        type_to_field = {
+            AnnotationType.DATE.value: fields.DateField(),
+            AnnotationType.STRING.value: fields.CharField(),
+            AnnotationType.INTEGER.value: fields.IntegerField(),
+            AnnotationType.DECIMAL.value: fields.FloatField(),
+        }
+        allowed_lookups = {
+            AnnotationType.DATE.value: DATE_LOOKUPS,
+            AnnotationType.STRING.value: TEXT_LOOKUPS,
+            AnnotationType.INTEGER.value: NUMBER_LOOKUPS,
+            AnnotationType.DECIMAL.value: NUMBER_LOOKUPS,
+        }
+        post_processing_map = {
+            AnnotationType.DATE.value: lambda value: value.isoformat()
+        }
+
+        match = re.match(
+            r"^(?P<field_id>\d+)(__(?P<lookup>\w+))?:(?P<value>.+)$", value
+        )
+        if not match:
+            raise ValidationError(f"Invalid annotation filter format '{value}'.")
+
+        field_id = match.group("field_id")
+        lookup_type = match.group("lookup") or "exact"
+        string_value = match.group("value")
+        field_type = (
+            AnnotationField.objects.filter(id=field_id)
+            .values_list("type", flat=True)
+            .get()
+        )
+
+        if lookup_type not in allowed_lookups[field_type]:
+            raise ValueError(
+                f"Lookup '{lookup_type}' not supported for the field '{field_id}'."
+            )
+
+        post_processing = post_processing_map.get(field_type, lambda x: x)
```

Comment:
Use `None` as default and execute the post processing only when it is defined.

Reply:
Fixed.